### ISSUE TITLE
GROUP BY over _shard_num doesn't work in distributed tables.

### ISSUE THREAD

**Issue by CurtizJ** [ClickHouse/ClickHouse#12601]
```
SELECT 
    _shard_num,
    count()
FROM remote('127.0.0.{1,2}', system.one)
GROUP BY _shard_num

↙ Progress: 0.00 rows, 0.00 B (0.00 rows/s., 0.00 B/s.) 
Received exception from server (version 20.7.1):
Code: 10. DB::Exception: Received from localhost:9000. DB::Exception: Not found column _shard_num in block. There are only columns: count(): While executing Remote. 
```

**Comment by CurtizJ** [comment]
I found a workaround:

```
WITH materialize(_shard_num) AS shard_num
SELECT 
    shard_num,
    count()
FROM remote('127.0.0.{1,2}', system.one)
GROUP BY shard_num

┌─shard_num─┬─count()─┐
│         2 │       1 │
│         1 │       1 │
└───────────┴─────────┘

2 rows in set. Elapsed: 0.003 sec.
```

**Comment by azat** [comment]
#16947 took too long in finding it path in upstream and now those patches does not work anymore, and not only #16947 does not work, but materialize trick does not helps either (for upstream):

```sql
select _shard_num, * from remote('127.{1,2}', system.one) limit 1 by dummy; -- fails
select materialize(_shard_num), * from remote('127.{1,2}', system.one) limit 1 by dummy; -- fails too
```

**Comment by azat** [comment]
Fixed in #24252

**Comment by azat** [comment]
@alexey-milovidov can you please close this?
